packages/idempotency/src/persistence/RedisPersistenceLayer.ts

```diff
@@ -0,0 +1,368 @@
+        | number
+        | undefined,
+      responseData: item[this.#dataAttr],
+      payloadHash: item[this.#validationKeyAttr] as string | undefined,
+    });
+  }
+
+  protected async _updateRecord(record: IdempotencyRecord): Promise<void> {
+    const item: Record<string, unknown> = {
+      [this.#statusAttr]: record.getStatus(),
+      [this.#expiryAttr]: record.expiryTimestamp,
+      [this.#dataAttr]: record.responseData,
+    };
+
+    const encodedItem = JSON.stringify(item);
+    const ttl = this.#getExpirySeconds(record.expiryTimestamp);
+    // Need to set ttl again, if we don't set `EX` here the record will not have a ttl
+    await this.#client.set(record.idempotencyKey, encodedItem, {
+      EX: ttl,
+    });
+  }
+
+  /**
+   * Determines if the provided Redis client is a default Redis client.
+   *
+   * This method checks if the provided client is an instance of the default Redis client
+   * by verifying the presence of the 'isOpen' property, which is specific to
+   * RedisClientType or RedisClusterType from the @redis/client package.
+   *
+   * @param client - The Redis client to check
+   */
+  #isDefaultRedisClient(
+    client: RedisClientProtocol | RedisClientType | RedisClusterType
+  ): client is RedisClientType | RedisClusterType {
+    return 'isOpen' in client;
+  }
+
+  /**
+   * Put a record in the persistence store with a status of "INPROGRESS".
+   * The method guards against concurrent execution by using Redis' conditional write operations.
+   */
+  async #putInProgressRecord(record: IdempotencyRecord): Promise<void> {
+    const item: Record<string, unknown> = {
+      [this.#statusAttr]: record.getStatus(),
+      [this.#expiryAttr]: record.expiryTimestamp,
+    };
+
+    if (record.inProgressExpiryTimestamp !== undefined) {
+      item[this.#inProgressExpiryAttr] = record.inProgressExpiryTimestamp;
+    }
+
+    if (this.isPayloadValidationEnabled() && record.payloadHash !== undefined) {
+      item[this.#validationKeyAttr] = record.payloadHash;
+    }
+
+    const encodedItem = JSON.stringify(item);
+    const ttl = this.#getExpirySeconds(record.expiryTimestamp);
+
+    try {
+      /**
+       * |     LOCKED     |         RETRY if status = "INPROGRESS"                |     RETRY
+       * |----------------|-------------------------------------------------------|-------------> .... (time)
+       * |             Lambda                                              Idempotency Record
+       * |             Timeout                                                 Timeout
+       * |       (in_progress_expiry)                                          (expiry)
+       *
+       * Conditions to successfully save a record:
+       *
+       * The idempotency key does not exist:
+       *   - first time that this invocation key is used
+       *   - previous invocation with the same key was deleted due to TTL
+       *   - SET see https://redis.io/commands/set/
+       */
+
+      console.debug(
```

Comment:
Let's remove all instances of `console.debug` for now - I know they're present in the Python implementation but we haven't added them in this version yet.

The reason is that right now there's no way to turn them off, so we avoided them and left only warnings.

In the medium term (perhaps right after this PR) we could do the same thing that we did for Metrics and the new Event Handler, and accept a `logger` class instance - so that the customer controls the level.

But for now better to remove it.